File: src/my_project/easy_problems/from201to250/longest_nice_substring.py
from typing import List 

class Solution:
    def longestNiceSubstring(self, s: str) -> str:
        def checkNice(s:str)-> bool:
            dict_answer = dict()
            answer = True
            for letter in s:
                dict_answer[letter] = 0
                if ord(letter) <= 90 and ord(letter) >= 65:
                    dict_answer[chr(ord(letter) + 32)] = 0
                else: 
                    dict_answer[chr(ord(letter) - 32)] = 0
                    
            for letter in s: 
                dict_answer[letter] += 1
                
            for letter in s:
                if ord(letter) <= 90 and ord(letter) >= 65:
                    if dict_answer[chr(ord(letter) + 32)] == 0:
                        answer = False 
                        return answer
                else: 
                    if dict_answer[chr(ord(letter) - 32)] == 0:
                        answer = False 
                        return answer       
                    
            return answer 
        
        len_s = len(s)
        max = -1
        answer: str = ""
        
        for i in range(len_s):
            for j in range(i+1, len_s):
                if checkNice(s[i:j+1]) and (len(s[i:j+1]) > max):
                    answer = s[i:j+1]
                    max = len(s[i:j+1])    
        return answer 
    
    
solution = Solution()

print(solution.longestNiceSubstring(s = "YazaAay"))

print(solution.longestNiceSubstring(s = "Bb"))

print(solution.longestNiceSubstring( s = "c"))

print(solution.longestNiceSubstring( s = "dDzeE"))


    

File: my_project_typescript/easy_problems_typescript/from101to150/longestNiceSubstring.ts
function longestNiceSubstring(s: string): string {

    function checkNice(s: string): boolean{
        
        let dictAnswer: any = {}
        let answer: boolean = true
        let lenS: number = s.length

        for (let i = 0; i <lenS; i++){
            dictAnswer[s[i]] = 0
            if (s[i].charCodeAt(0) <= 90 && s[i].charCodeAt(0)>= 65) {
                dictAnswer[String.fromCharCode(s[i].charCodeAt(0)+ 32 ) ] = 0
            }
            else {
                dictAnswer[String.fromCharCode(s[i].charCodeAt(0) - 32) ] = 0
            }
        }

        for (let i = 0; i < lenS; i++){
            dictAnswer[s[i]] += 1
        }

        for (let i = 0; i < lenS; i++){
            if (s[i].charCodeAt(0) <= 90 && s[i].charCodeAt(0)>= 65) {
                 if (dictAnswer[String.fromCharCode(s[i].charCodeAt(0)+ 32 ) ] == 0){
                     answer = false 
                     return answer 
                 }
            }
            else {
                if (dictAnswer[String.fromCharCode(s[i].charCodeAt(0) - 32) ] == 0){
                    answer = false 
                    return answer 
                }
            }
        }
        return answer 
    }

    let lenS: number = s.length
    let max: number = -1
    let answer: string = ''

    for (let i = 0; i < lenS; i++){
        for (let j = 0; j < lenS; j++){
            if (checkNice(s.slice(i, j+1)) && ((j - i + 1) > max)){
                answer = s.slice(i, j+1)
                max  = j - i + 1
            }
        }
    }

    return answer
};

console.log(longestNiceSubstring("YazaAay"))

console.log(longestNiceSubstring("Bb"))

console.log(longestNiceSubstring("c"))

console.log(longestNiceSubstring("dDzeE"))

